online-bulk-load/src/main/scala/org/tikv/bulkload/RawKVBulkLoader.scala

```diff
@@ -116,40 +116,44 @@ class RawKVBulkLoader(tiConf: TiConfiguration, sparkConf: SparkConf) extends Ser
   }
 
   private def writeAndIngest(iterator: Iterator[(Array[Byte], Array[Byte])], partitioner: RegionPartitioner): Unit = {
-    val (itor1, tiro2) = iterator.duplicate
-
-    var minKey: Key = Key.MAX
-    var maxKey: Key = Key.MIN
-    var region: TiRegion = null
-    var key: Key = null
-    val dataSize = itor1.map { itor =>
-      key = Key.toRawKey(itor._1)
-
-      if (region == null) {
-        region = partitioner.getRegion(key)
+    val sortedList = new util.ArrayList[Pair[ByteString, ByteString]](iterator.toList.map(pair => Pair.create(ByteString.copyFrom(pair._1), ByteString.copyFrom(pair._2))).asJava)
+    sortedList.sort(new Comparator[Pair[ByteString, ByteString]]() {
+      override def compare(o1: Pair[ByteString, ByteString], o2: Pair[ByteString, ByteString]): Int = {
+        Key.toRawKey(o1.first).compareTo(Key.toRawKey(o2.first))
       }
+    })
 
-      if (key.compareTo(minKey) < 0) {
-        minKey = key
-      }
-      if (key.compareTo(maxKey) > 0) {
-        maxKey = key
-      }
-    }.size
-
+    val dataSize = sortedList.size()
     if (dataSize > 0) {
+      val minKey: Key = Key.toRawKey(sortedList.get(0).first)
+      val maxKey: Key = Key.toRawKey(sortedList.get(sortedList.size() - 1).first)
+      val region: TiRegion = partitioner.getRegion(minKey)
+
+      logger.info(
+        s"""
+           |dataSize=$dataSize
+           |minKey=${minKey.toByteString.toStringUtf8}
+           |maxKey=${maxKey.toByteString.toStringUtf8}
+           |region=$region
+           |""".stripMargin)
+
       if (region == null) {
-        logger.warn("region == null, skip ingest this partition")
+        throw new Exception("region == null")
       } else {
         val uuid = genUUID()
 
         logger.info(s"start to ingest this partition ${util.Arrays.toString(uuid)}")
-        val pairsIterator = tiro2.map { keyValue =>
-          new Pair[ByteString, ByteString](ByteString.copyFrom(keyValue._1), ByteString.copyFrom(keyValue._2))
-        }.toList.sortWith((a, b) => Key.toRawKey(a.first).compareTo(Key.toRawKey(b.first)) < 0).toIterator.asJava
-        val tiSession = TiSession.create(tiConf)
+        var tiSession: TiSession = null
+        while(tiSession == null) {
+          try {
+            tiSession = TiSession.create(tiConf)
+          } catch {
+            case e: Throwable => logger.warn("create tiSession failed!", e)
```

Comment:
If create TiSession something error, the process will never exit

Reply:
sure, i will fix it

Reply:
fixed. PTAL